Add Root sampler, Trace ID ratio based sampler options

kaylareopelle opened the pull request:
* Adds the trace id ratio based sampling options, along with ratio configs for each of those options
  * This sampling matches the OpenTelemetry Ruby SDK sampler and not the recently defined algorithm that will be used for the Probability Sampler
* Adds the `adaptive` option to all DT samplers
* Adds the ability to configure the root sampling decision to the same options as the remote parent samplers: `default`, `adaptive`, `trace_id_ratio_based`, `always_on`, `always_off`

Closes #3284 
Extends #3307

kaylareopelle committed:
Add root config, add adaptive to option list

kaylareopelle committed:
Remove unused method enforce_fallback

We have the allowlist functionality now that is faster.

kaylareopelle committed:
Enforce sampler config values with allowlist

kaylareopelle committed:
Add support for four sampler options

default, adaptive, always_on, always_off are now available for use when
a trace is root, remote_parent_sampled or remote_parent_not_sampled.

default is first in the case list because it is the most common. I
predict adaptive will be far less common than always_on or always_off,
so I put it last for a slight performance benefit.

kaylareopelle committed:
Add trace_id_ratio_based support to root sampling

kaylareopelle committed:
WIP trace ratio in DT and TC

kaylareopelle committed:
Brute force trace id ratio sampler DT

kaylareopelle committed:
Update default value for ratio to be nil

kaylareopelle committed:
Merge branch 'dev' into trace-ratio

kaylareopelle committed:
Refactor set_priority_and_sampled to use configs

kaylareopelle committed:
Add note to update priority algo for trace ratio

kaylareopelle committed:
Refactor config validation

Move validation to a separate module and use a transform to call those
methods.

Remove ratio validation checks and fallbacks in Transaction,
TraceContext, and DistributedTracing.

kaylareopelle committed:
Remove unique priority method for NR header

Since both DistributedTracing#set_priority_and_sampled_newrelic_header
and TraceContext#set_priority_and_sampled are included in the
Transaction class and the methods ended up being identical, we can
define it in only one spot.

kaylareopelle committed:
Use adaptive_priority for trace_id_ratio_based

Setting it to all or nothing isn't compatible with the spec

kaylareopelle committed:
Add trace_ratio_sampled?

Now the trace ratio sampling logic lives in one method

kaylareopelle committed:
Call methods from Transaction instance

kaylareopelle committed:
Reduce, refactor, rubocop

kaylareopelle committed:
Update trace_ratio_sampled? comment

kaylareopelle committed:
Add tests

kaylareopelle committed:
Improve performance of trace_ratio_sampled?

OpenTelemetry stores their trace IDs as binary strings:
"\xAB\xCD\xEF\x01#Eg\x89"
We keep ours as alphanumeric strings (ex: "abcdef0123456789")

Rather than coerce every trace ID into a binary string, we can use
Integer to translate our guids into the same value as the
unpacked binary by translating them into base 16.

kaylareopelle committed:
Add changelog entry

kaylareopelle committed:
Cleanup, rubocop

kaylareopelle commented on lib/new_relic/agent/configuration/default_source.rb:
```diff
@@ -281,16 +282,6 @@ def self.convert_to_constant_list(raw_value)
           constants
         end
-
-        def self.enforce_fallback(allowed_values: nil, fallback: nil)
```
This is unused code. We have an allowlist feature now with a fallback function that's more performant.

kaylareopelle commented on lib/new_relic/agent/configuration/default_source.rb:
```diff
@@ -1463,19 +1454,68 @@ def self.enforce_fallback(allowed_values: nil, fallback: nil)
+          :default => 'default',
+          :public => true,
+          :type => String,
+          :allowed_from_server => false,
```
Don't know if these should or shouldn't be allowed

kaylareopelle replied:
@tannalynn - 📰 None of these configs are expected to be allowed from the server.

hannahramadan commented on lib/new_relic/agent/configuration/default_source.rb:
```diff
@@ -1463,19 +1454,68 @@ def self.enforce_fallback(allowed_values: nil, fallback: nil)
+          :public => true,
+          :type => String,
+          :allowed_from_server => false,
+          :allowlist => %w[default adaptive always_on always_off trace_id_ratio_based],
```
What is behind the `default` value being `default` instead of just `adaptive`? Did we need to have `default` as a kept option per like the spec or something

hannahramadan commented on lib/new_relic/agent/configuration/default_source.rb:
```diff
@@ -1463,19 +1454,68 @@ def self.enforce_fallback(allowed_values: nil, fallback: nil)
-          :allowed_from_server => true,
-          :description => 'This setting controls the behavior of transaction sampling when a remote parent is not sampled and the trace flag is not set in the traceparent. Available values are `default`, `always_on`, and `always_off`.'
+          :allowed_from_server => false,
+          :allowlist => %w[default adaptive always_on always_off trace_id_ratio_based],
```
I like this allowlist feature

kaylareopelle replied:
me too! was happy to find it. a wonderful james feature 🍵 🦊

hannahramadan commented on lib/new_relic/agent/transaction.rb:
```diff
@@ -307,8 +328,24 @@ def trace_id=(value)
         @trace_id = value
       end
 
+      # This may be better in the adaptive sampler?
```
I think I agree! It feels more intuitive to live in the adaptive sampler class

hannahramadan commented on lib/new_relic/agent/transaction/trace_context.rb:
```diff
@@ -171,14 +179,20 @@ def use_nr_tracestate_sampled(payload)
-          if config == 'always_on'
+        def set_priority_and_sampled(sampler, ratio, payload)
+          case sampler
+          when 'default'
```
I would take a comment! In the context of this entire PR it makes sense, but I could see looking at this in isolation more confusing

kaylareopelle replied:
Sounds good! Added in 85cfd5dcc

kaylareopelle committed:
Add comments about case statement order

kaylareopelle committed:
Apply suggestion from @hannahramadan

Co-authored-by: Hannah Ramadan <76922290+hannahramadan@users.noreply.github.com>

kaylareopelle committed:
Merge branch 'trace-ratio-with-config-refactor' of https://github.com/newrelic/newrelic-ruby-agent into trace-ratio-with-config-refactor

kaylareopelle committed:
Rename adaptive_priority to default_priority

The method is used to determine priorities outside of the adaptive
sampler and is the legacy, default priority

kaylareopelle committed:
Merge branch 'dev' into trace-ratio-with-config-refactor

kaylareopelle committed:
Merge branch 'dev' into trace-ratio-with-config-refactor

kaylareopelle committed:
Merge branch 'trace-ratio-with-config-refactor' of https://github.com/newrelic/newrelic-ruby-agent into trace-ratio-with-config-refactor
